core/rawdb/chain_freezer.go

```diff
@@ -114,60 +122,39 @@ func (f *chainFreezer) freeze(db ethdb.KeyValueStore) {
 				return
 			}
 		}
-		// Retrieve the freezing threshold.
-		hash := ReadHeadBlockHash(nfdb)
-		if hash == (common.Hash{}) {
-			log.Debug("Current full block hash unavailable") // new chain, empty database
-			backoff = true
+		finalNumber, finalHash, err := f.readFinalized(nfdb)
+		if err != nil {
+			backoff = true // chain is not finalized yet
+			log.Debug("Finalized block is not available yet")
 			continue
 		}
-		number := ReadHeaderNumber(nfdb, hash)
-		threshold := f.threshold.Load()
 		frozen := f.frozen.Load()
-		switch {
-		case number == nil:
-			log.Error("Current full block number unavailable", "hash", hash)
-			backoff = true
-			continue
 
-		case *number < threshold:
-			log.Debug("Current full block not old enough to freeze", "number", *number, "hash", hash, "delay", threshold)
-			backoff = true
-			continue
-
-		case *number-threshold <= frozen:
-			log.Debug("Ancient blocks frozen already", "number", *number, "hash", hash, "frozen", frozen)
-			backoff = true
-			continue
-		}
-		head := ReadHeader(nfdb, hash, *number)
-		if head == nil {
-			log.Error("Current full block unavailable", "number", *number, "hash", hash)
+		// Short circuit if finalized blocks are already frozen.
+		if frozen != 0 && frozen-1 >= finalNumber {
 			backoff = true
+			log.Debug("Ancient blocks frozen already", "number", finalNumber, "hash", finalHash, "frozen", frozen)
 			continue
 		}
-
 		// Seems we have data ready to be frozen, process in usable batches
 		var (
-			start    = time.Now()
-			first, _ = f.Ancients()
-			limit    = *number - threshold
+			start = time.Now()
+			first = frozen
+			limit = finalNumber
 		)
-		if limit-first > freezerBatchLimit {
-			limit = first + freezerBatchLimit
+		if limit-first+1 > freezerBatchLimit {
+			limit = freezerBatchLimit + first - 1
```

Comment:
What is the reason for the `+1` and `- 1` ?

Reply:
Because in the `f.freezeRange(nfdb, first, limit)`, we will freeze the blocks in range [first, limit], limit is also included. The frozen block number is `limit-first+1`.

The `freezerBatchLimit` is the cap of the block number, therefore the limit is capped to `freezerBatchLimit + first - 1` if too many blocks in a single run.